### PR TITLE
GH-725: Added support for multi-gpu in language model training

### PR DESCRIPTION
**Pull request by jasminesjlee**
Hello! I added some code to make training language models compatible with multiple GPUs if they are available. I spent some time doing this so that I could pretrain some models more quickly, but thought I would share it with you as well in case it will help. I saw that there were many issues addressing this but that it has not been implemented yet.

### CODE REVIEW

**Comment by kashif** [comment]
dear @jasminesjlee thank you for the PR. It seems you are using the `DataParallel` which will only place the model on multiple gpus, however to train with multiple gpu you will need to use `DistributedDataParallel`.

**Comment by jasminesjlee** [comment]
Hello! Can you clarify why? My code also calculates the predictions/loss in parallel.

**Comment by kashif** [comment]
@jasminesjlee Yes you are right, `DataParallel` will chunk the data across the batch dimension and sum the gradients in the backward pass. However i would have thought you would only need to just wrap the `self.model` with it as the approach you have seems a bit too involved... is it beacuse there is no `forward` method in the `FlairModel` ?

**Comment by jasminesjlee** [comment]
There is a `forward` method in the `LanguageModel` class, but it outputs predictions. I did not want to change this, but I still wanted to calculate loss in parallel, so I had to define a class `FlairLoss` that calculates loss in the `forward` method. It would also be possible for me to remove the `FlairLoss` class and instead use the `forward` method of the `LanguageModel` class to output loss, if that is what you prefer!

**Comment by jbesomi** [comment]
Hey,

I believe supporting multi-GPU training is quite important for Flair as it would drastically reduce training time. I plan to use Flair quite extensively in the next few months. I have four GPU (RTX 2080) and I would like to use all of them during training. That said, I would like to help you support multi GPUs, starting from LMs and later on adding support for all task-specific training.

@alanakbik, in #936 you said that "we tend to favor horovod [rather than DataParallel] since this is less invasive to the rest of the code and has more functionality". From my understanding, and by looking at the official [pytorch multi-GPU tutorial](https://pytorch.org/tutorials/beginner/blitz/data_parallel_tutorial.html), for multi-gpu training it suffices to wrap the PyTorch model into a `nn. DataParallel` class. Assuming that works, I believe this is much less invasive than adding an external dependency such as `horovod` (considering many factors, for instance, code maintainability).

@jasminesjlee I don't get what you just wrote in your last comment. Can you please try to explain a bit better? Wouldn't be enough to do something like shown below? Thanks!

(`language_model_trainer.py `)

```
class LanguageModelTrainer:
    def __init__(..., model, ...):
    
        if torch.cuda.device_count() > 1:
           self.model: LanguageModel =  nn.DataParallel(model)
```
        
@alanakbik it would be great to receive support from someone more expert that had to deal with multi-processing in the past. May you please reopen this issue and add a `help wanted` label?

Thanks,
Jonathan

**Comment by djstrong** [comment]
We tried DataParallel but it was not so straightforward. Finally, we only parallelized LSTM: `self.rnn = torch.nn.DataParallel(nn.LSTM(embedding_size, hidden_size, nlayers), dim=1)`.

**Comment by jbesomi** [comment]
Hey @djstrong, thank you for your comment. Can you please add more information? I would love to know more. For instance, who is "we" and why is it not so `straightforward `? Which kind of issues did you get? How big was the gain with `nn.DataParallel`? Moreover, why are you specifying `dim=1` (default is 0)?

**Comment by alanakbik** [comment]
Hello @jbaiter - as it happens, @kishaloyhalder has very recently created a (I believe) working version of the DataParallel solution that you describe. If it turns out to be stable, it would be great to add this to Flair.

**Comment by djstrong** [comment]
@jbesomi It was 3 years ago and I don't remember: https://github.com/djstrong/flair/commit/9683254943e019369d5082d90c8923eb801974fb
The speed gain was not so good, about 50% for two cards.

**Comment by alanakbik** [comment]
I believe @kishaloyhalder got more of an improvement, but it was a very GPU-heady task.

**Comment by jbesomi** [comment]
@alanakbik do you know where we can find @kishaloyhalder multi-GPU code approach?

**Comment by kishaloyhalder** [comment]
Hi @jbesomi , @djstrong 

I have been able to train a `TextClassifier` model with multiple GPUs with Flair using pytorch's DataParallel. However, it is not is a shape to create a PR yet. If your use case overlaps with this, I would be happy to share the code (via a tarball or a new branch maybe) with you.

If you are interested in how it works: Flair uses a custom object called `Sentence` to pass around between trainer and the model. Pytorch's DataParallel assumes all the data would be in [tensor format](https://github.com/pytorch/pytorch/blob/04e7891aab09bda348964c4ae62db66265141936/torch/nn/parallel/scatter_gather.py#L19). It distributes them evenly across all the GPU cards and gathers the losses back. However, in Flair since the data object is not tensor it does not work well with the off-the-shelf DataParallel. I had overridden the `scatter` method in the DataParallel to be compatible with `Sentence` objects.

with regards,
Kishaloy

**Comment by jbesomi** [comment]
Hi @kishaloyhalder,

Thanks, I see. I would be more than interested to have a look at the code and see if it might come in handy to help me implement multi-GPU support for LMs training. For sharing, both tarballs or new branch works. Out of curiosity, how big was the speed gain? Thanks!
